README
define schemas as a hash, and validate hashes!

    require 'hashidator'

    schema  = {:name => String, :age => Integer}
    valid   = {:name => "Harry", :age => 20}
    invalid = {:name => 1234, :age => "twenty"}

    h = Hashidator.new(schema)
    h.validate(valid)   #=> true
    h.validate(invalid) #=> false

see examples/basic.rb for full coverage. it knows
classes, ranges, booleans, proc and duck typing. yay!

to release this gem:
 * edit lib/hashidator/version.rb
 * edit CHANGES
 * git commit -am "Release VERSION"
 * rake release

(c) 2009 harry vangberg <harry@vangberg.name>,
         peter suschlik <peter@suschlik.de>

distributed under the mit license. google it.


CHANGES
= 0.4.1 / 2016-03-07

  * Modernize project (bundler, minitest). [Peter Suschlik]

= 0.4.0 / 2011-06-07

  * Note: This was pushed to rubygems on 2016-03-07

  * Better support for subclassing Hashidator. [Peter Suschlik]

  * Custom validator classes need to implement #===. [Peter Suschlik]

    Optional value example:

      class OpionalValue
        def initialize(schema)
          @schema = schema
        end

        def ===(value)
          value.nil? || @schema
        end
      end

      schema = Hashidator.new(:name => String, :hobby => OpionalValue.new(String))
      schema.validate(:name => "Bob", :hobby => "Cipher") # => true
      schema.validate(:name => "Bob")                     # => true, even without a hobby
      schema.validate(:name => "Bob", :hobby => 23)       # => falseo

  * Add string validator [Peter Suschlik]

= 0.3.1 / 2010-10-05

  * Bug fix: Failing on nil as input [Peter Suschlik]
  * Bug fix: Failing on non-array input to array schema [Peter Suschlik]

= 0.3 / 2010-02-03

  * Validate with procs. [Peter Suschlik]
  
    Proc returns a boolean or another validator.
    Now you can write custom validators like:

      schema = Hashidator.new(:ary => proc {|v| v.size == 3 })
      schema.validate(:ary => [1,2,3]) # => true
      schema.validate(:ary => [1,2])   # => false

= 0.2 / 2009-12-19

  * Validate regexp:  :foo => /match this/

  * Validate hashes in arrays:  [{:foo => String}]  [Peter Suschlik]

= 0.1 / 2009-12-18

  * Release 0.1


test/helper.rb
require 'minitest/autorun'
require 'hashidator'

class Minitest::Test
  def assert_true(subject)
    assert_equal true, subject
  end

  def assert_false(subject)
    assert_equal false, subject
  end
end


test/test_hashidator.rb
require 'helper'

class TestHashidator < Minitest::Test
  def h(schema, input)
    Hashidator.validate(schema, input)
  end

  def test_validate_integer
    assert_true h({:age => Integer}, {:age => 20})
  end

  def test_invalidate_integer
    assert_false h({:age => Integer}, {:age => "twenty"})
  end

  def test_validate_string
    assert_true h({:name => String}, {:name => "Harry"})
  end

  def test_invalidate_string
    assert_false h({:name => String}, {:name => 44667})
  end

  def test_validate_in_range
    assert_true h({:age => (10..20)}, {:age => 10})
    assert_true h({:age => (10..20)}, {:age => 15})
    assert_true h({:age => (10..20)}, {:age => 20})
  end

  def test_invalidate_in_range
    assert_false h({:age => (10..20)}, {:age => 9})
    assert_false h({:age => (10..20)}, {:age => 21})
  end

  def test_validate_array_members
    assert_true h({:children => [String]}, {:children => ["Sue", "Mike"]})
  end

  def test_validate_empty_array_members
    assert_true h({:children => [String]}, {:children => []})
  end

  def test_invalidate_array_members
    assert_false h({:children => [String]}, {:children => ["Sue", 1234]})
  end

  def test_invalidate_array_members_with_non_array_values
    assert_false h({:ary => [String]}, {:ary => nil})
    assert_false h({:ary => [String]}, {:ary => :not_an_array})
    assert_false h({:ary => [String]}, {:ary => {:i_am => "a hash"}})
  end

  def test_validate_boolean
    assert_true h({:admin => Boolean}, {:admin => true})
  end

  def test_invalidate_boolean
    assert_false h({:admin => Boolean}, {:admin => 1234})
  end

  def test_validate_regexp
    assert_true h({:uri => /^http:/}, {:uri => "http://example.com"})
  end

  def test_invalidate_regexp
    assert_false h({:uri => /^http:/}, {:uri => "john coltrane"})
    assert_false h({:uri => /^http:/}, {:uri => nil})
    assert_false h({:uri => /^http:/}, {:uri => :symbol})
  end

  def test_validate_exact_values
    assert_true h({:result => "ok"}, {:result => "ok"})
    assert_true h({:bool => true}, {:bool => true})
  end

  def test_invalidate_exact_values
    assert_false h({:result => "ok"}, {:result => "error"})
    assert_false h({:result => "ok"}, {:result => nil})
    assert_false h({:bool => false}, {:bool => true})
  end

  def test_validate_respond_to
    assert_true h({:name => :to_s}, {:name => "Harry"})
  end

  def test_invalidate_respond_to
    assert_false h({:name => :non_existing_method}, {:name => "Harry"})
  end

  def test_validate_array_members_respond_to
    assert_true h(
      {:names => [:to_s]},
      {:names => ["Harry", "Damone"]}
    )
  end

  def test_invalidate_array_members_respond_to
    assert_false h(
      {:name => [:non_existing_method]},
      {:name => ["Harry", "Damone"]}
    )
  end

  def test_validate_array_members_range
    assert_true h(
      {:names => [(1..9)]},
      {:names => [1, 9]}
    )
  end

  def test_validate_proc_with_boolean_result
    assert_true h(
      {:methods => proc {|v| v.all? {|e| e.is_a?(String) || e.is_a?(Symbol) } }},
      {:methods => ["to_s", :to_h]}
    )
  end

  def test_invalidate_proc_with_boolean_result
    assert_false h(
      {:methods => proc {|v| v.all? {|e| e.is_a?(String) || e.is_a?(Symbol) } }},
      {:methods => ["to_s", 123]}
    )
  end

  def test_validate_proc_with_cascaded_result
    schema = {:methods => proc {|v| v.first.is_a?(Integer) ? [Integer] : [String] } }
    assert_true h(schema, {:methods => [1, 2]})
    assert_true h(schema, {:methods => ["Harry", "Damone"]})
  end

  def test_invalidate_proc_with_cascaded_result
    schema = {:methods => proc {|v| v.first.is_a?(Integer) ? [Integer] : [String] } }
    assert_false h(schema, {:methods => [1, "Damone"]})
    assert_false h(schema, {:methods => ["Harry", 2]})
  end

  def test_validation_with_equality
    optional = Class.new do
      def initialize(schema)
        @schema = schema
      end

      def ===(other)
        other.nil? || @schema
      end
    end

    schema = {:phone => optional.new(String)}
    assert_true h(schema, {:phone => "+49 2305 4711"})
    assert_true h(schema, {:phone => nil})
    assert_true h(schema, {})
    assert_false h(schema, {:phone => false})
    assert_false h(schema, {:phone => 1234567})
  end

  def test_validate_nested
    schema = {:name => {:first => String, :last => String}}
    assert_true h(schema, {:name => {:first => "Mike", :last => "Damone"}})
  end

  def test_invalidate_nested
    schema = {:name => {:first => String, :last => String}}
    assert_false h(schema, {:name => {:first => "Mike", :last => 1234}})
  end

  def test_validate_deep_nested
    assert_true h(
      {:people => [{ :name => :to_s, :age => (18..40)}]},
      {:people => [{ :name => "Ann", :age => 19 }, { :name => "Bob", :age => 23 }]}
    )
  end

  def test_invalidate_deep_nested
    assert_false h(
      {:people => [{ :name => :to_s, :age => (18..40)}]},
      {:people => [{ :name => "Ann", :age => 15 }, { :name => "Bob", :age => 23 }]}
    )
  end

  def test_deep_nested_validation_uses_subclass
    h = Class.new(Hashidator) do
      def validate(input)
        throw :halt, "#{input[:value]} HAMMERZEIT" if input[:value] == 'STOP'
        super
      end
    end

    caught = catch(:halt) do
      h.validate(
        {:deep => { :value => String}},
        {:deep => { :value => 'STOP'}})
    end
    assert_equal "STOP HAMMERZEIT", caught
  end

  def test_invalidate_for_nil_input
    schema = {:user => { :name => String }}
    assert_false h(schema, nil)
    assert_false h(schema, {})
    assert_false h(schema, {:user => nil})
  end
end


lib/hashidator.rb
class Hashidator
  autoload :VERSION, 'hashidator/version'

  def self.validate(schema, input)
    new(schema).validate(input)
  end

  attr_accessor :schema

  def initialize(schema)
    @schema = schema
  end

  def validate(input)
    input ||= {}
    schema.all? {|key, validator|
      validate_value(validator, input[key])
    }
  end

  private

  def validate_value(validator, value)
    case validator
    when Array
      value.respond_to?(:all?) && value.all? {|x| validate_value(validator[0], x)}
    when Hash
      value.respond_to?(:[]) && validator.all? {|k, v| validate_value(v, value[k]) }
    when Symbol
      value.respond_to? validator
    else
      result = validator == value
      result = validator === value unless result
      result = validate_value(result, value) unless Boolean === result
      result
    end
  end
end

module Boolean
end

class TrueClass
  include Boolean
end

class FalseClass
  include Boolean
end

class Proc
  alias :=== :call if RUBY_VERSION < "1.9"
end


examples/basic.rb
require 'hashidator'

class OptionalValue
  def initialize(schema)
    @schema = schema
  end

  def ===(value)
    value.nil? || @schema
  end
end

schema = {
  :id     => Integer,         # Is integer?
  :name   => String,
  :age    => (13..99),        # Within range?
  :admin  => Boolean,
  :mails  => [String],        # Array consisting of strings?
  :other  => {
    :country_code => :to_i,   # Ducktyping!
    :country      => String,
    :random       => /foo/    # Regular expressions
  },
  :result => "ok",
  :bool   => true,
  :hotchicks => OptionalValue.new([{ :name => String, :age => (18..30) }]),
  :validate_array       => [{:name => String}],
  :validate_array_size  => proc {|v| v.size == 3 }
}

valid_input = {
  :id     => 123,
  :name   => "Harry",
  :age    => 21,
  :admin  => true,
  :mails  => ["foo@example.com", "bar@example.com"],
  :other  => {
    :country      => "Denmark",
    :country_code => 12,
    :random       => "foobar"
  },
  :result => "ok",
  :bool => true,
  :hotchicks => [
    { :name => "Nura",  :age => 26 },
    { :name => "Sarah", :age => 30 }
  ],
  :validate_array => [
    {:name => "John"},
    {:name => "Coltrane"}
  ],
  :validate_array_size => [1,2,3]
}

invalid_input = {
  :id     => "whatevz",
  :name   => 42,
  :age    => 12,
  :admin  => :maybe,
  :mails  => ["foo@example.com", 1234],
  :other => {
    :country      => [1,2,3],
    :country_code => (1..2),
    :random       => "nothing here"
  },
  :result => "error",
  :bool => false,
  :hotchicks => [
    { :name => "Uma", :age => 42 }
  ],
  :validate_array => [
    {:name => "John"},
    {:name => 1234}
  ],
  :validate_array_size => [1,2]
}

h = Hashidator.new(schema)
puts h.validate(valid_input)     #=> true
puts "=" * 80
puts h.validate(invalid_input)   #=> false
